In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
lending_data_path = Path("Resources/lending_data.csv")
lending_df = pd.read_csv(lending_data_path)

# Review the DataFrame
# YOUR CODE HERE!
lending_df.head(10)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  
5                 1       20600            0  
6                 1       21100            0  
7                 0       15100            0  
8                 0       17400            0  
9                 0       18800            0

In [3]:
lending_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


In [4]:
lending_df.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [5]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_df["loan_status"]

# Separate the X variable, the features
# YOUR CODE HERE!
X = lending_df.drop(columns="loan_status")

In [6]:
# Review the y variable Series
# YOUR CODE HERE!
y.head(10)

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: loan_status, dtype: int64

In [7]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X.head(10)

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
5                 1       20600  
6                 1       21100  
7                 0       15100  
8                 0       17400  
9                 0       18800

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [11]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1,stratify=y)
X_train.shape

(58152, 7)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier


LogisticRegression(random_state=1)

In [13]:

# Fit the model using training data
# YOUR CODE HERE!
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

In [14]:
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.9926743706149401
Testing Data Score: 0.993912505158894


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [15]:
# Make a prediction using the testing data
# YOUR CODE HERE!
test_predictions = classifier.predict(X_test)
results = pd.DataFrame({"Prediction": test_predictions, "Actual": y_test})
results.head(20)

Prediction  Actual
36831           0       0
75818           1       1
36563           0       0
13237           0       0
43292           0       0
68423           0       0
37714           0       0
64870           0       0
47959           0       0
49              0       0
4010            0       0
44780           0       0
8147            0       0
51770           0       0
133             0       0
67757           0       0
15207           0       0
36131           0       0
70949           0       0
68035           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
test_matrix = confusion_matrix(y_test, test_predictions)
print(test_matrix)

[[18673    86]
 [   32   593]]


In [17]:
# Print the classification report for the model
# YOUR CODE HERE!
test_report = classification_report(y_test, test_predictions)
print(test_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.95      0.91       625

    accuracy                           0.99     19384
   macro avg       0.94      0.97      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The overall accuracy indicates that the model is highly effective for this dataset. Based on the metrics above, the logistic regression model is very effective at predicting both healthy and high-risk loans, with particularly strong performance in identifying healthy loans. The logistic regression model is performing exceptionally well in predicting healthy loans (0). Its performance is also very strong in identifying high-risk loans (1), albeit with slightly lower precision. The high recall for high-risk loans is particularly important in a credit risk context, as it means the model is effective at flagging most of the potential risky loans.

### Credit Risk Analysis Report

For 0 (Healthy Loan):
    
    *Precision: 1.00 -This means that when the model predicts a loan as healthy, it is correct 100% of the time.
    
    *Recall: 1.00 - This indicates that the model successfully identifies 100% of the actual healthy loans.

    *F1-Score: 1.00 - The F1-score, which is a balance between precision and recall, is perfect for healthy loans, indicating excellent model performance for this class.

For 1 (High-Risk Loan):
    
    *Precision: 0.87 - When the model predicts a loan as high-risk, it is correct 87% of the time.

    *Recall: 0.95 - The model identifies 95% of all actual high-risk loans correctly.

    *F1-Score: 0.91 - This is a strong F1-score, suggesting a good balance between precision and recall for high-risk loans.

    

---